### DISCUSSION TITLE
variables

### DISCUSSION THREAD

**Discussion by giscus** [sunface/rust-by-practice#181] [Book Comments]
# variables

Learning Rust By Practice, narrowing the gap between beginner and skilled-dev with challenging examples, exercises and projects.

https://zh.practice.rs/variables.html

**Comment by zengleyi** [comment]
mark一下

**Comment by codecow-s** [comment]
mark一下-2022-4-14

**Comment by Tailless-Monkey** [comment]
第一题感觉这样，更容易理解   println!("x is equal to {}", x);

**Reply by sunface** [reply]
好像、确实如此🤣 可以提个 PR :D

**Reply by ZYKWLJ** [reply]
hhhhh

**Comment by shijianzhong** [comment]
没看到答案呢

**Reply by shijianzhong** [reply]
fn main() {
    let x: i32 = 5; // uninitialized but using, ERROR !
    let _y: i32; 
    println!("{} is equal to 5", x); 
}

**Comment by Maplecc9** [comment]
打卡

**Comment by fireindark707** [comment]
完成，9/9

**Comment by RomanSeven7** [comment]
打开

**Comment by graydonli** [comment]
🔥 整个 rust-by-practice 所有题目，哪里有答案呢？

**Reply by Git4Mark** [reply]
https://github.com/sunface/rust-by-practice/tree/master/solutions

**Reply by graydonli** [reply]
thx!

**Comment by tao1991123** [comment]
mark

**Comment by makexiao113** [comment]
作业已完成。花了15分钟，感觉是比较基础的练习，对于rust的变量可变性，元组，数组，结构，作用域，变量的遮蔽性，还有所有权都有一定的考察，循序渐进，很有正反馈的练习作业

**Comment by ggchengyu** [comment]
未使用变量第二种方法。。。怎么说呢，是我好奇点了一下眼睛，看到shebang就懂了。。

**Comment by Neisseri** [comment]
请问第4题答案中为什么需要在"hello"后面加一个to_string()呢

**Reply by sweetmainW** [reply]
变量所有权的问题

**Reply by ghost** [reply]
因为答案的函数返回的值是String类型，而"hello"在Rust中是另一种类型：&str（本身是一个字符串切片）。所以，为了让返回值的类型和签名的返回类型一致，需要通过.to_string()方法将String类型转换为&str类型。

如果要返回&str的话，应该是这样的：

fn main() {
    println!("{}, world", define_x()); 
}

fn define_x<'a>() -> &'a str {
    let x = "hello";
    x
}

**Reply by Neisseri** [reply]
好的我懂了，谢谢两位

**Reply by oplancelot** [reply]
fn define_x<'a>() -> &'a str {} 这个函数格式比较难懂

ChatGPT
我理解你的困惑。让我更详细地解释一下这个函数的格式。

rust
Copy code
fn define_x<'a>() -> &'a str {}
在 Rust 中，这个函数的声明使用了泛型和生命周期参数，可能会让它看起来比较复杂。我会分步解释每个部分的含义：

fn: 这是 Rust 中定义函数的关键字。

define_x: 这是函数的名称。

<'a>: 这是一个泛型参数声明，'a 是一个生命周期标识符。生命周期参数用来指示变量或引用的有效范围。

(): 这是函数的参数列表，没有参数，所以是一个空括号。

->: 这表示函数返回类型的定义即将开始。

&'a str: 这是函数的返回类型。它表示一个生命周期为 'a 的字符串引用。在 Rust 中，&str 表示字符串切片，而 &'a str 表示具有生命周期 'a 的字符串切片。

综合起来，函数 define_x 声明了一个没有参数的函数，返回一个具有生命周期 'a 的字符串切片引用。这个生命周期参数 'a 意味着函数返回的引用必须在 'a 这个生命周期内保持有效。

这种函数声明的主要目的是确保函数返回的引用在调用它的地方能够正确地保持有效性，并且与调用函数的上下文生命周期相匹配。生命周期是 Rust 的一种安全性机制，用于避免悬垂引用和内存安全问题。

**Comment by zhouyiqi91** [comment]
pass

**Comment by wd2010** [comment]
这种学习方式确实很不错，实时的正反馈，感谢楼主

**Comment by kimchi-shi** [comment]
第4题第二个答案
```
fn main() {
    let x = define_x();
    println!("{:?}, world", x);
}

fn define_x() -> &'static str {
    let x = "hello";
    x
}
```
里面{:?} 和 &'static str 分别是什么意思啊？

**Reply by ghost** [reply]
（1）{:?} 是实现Debug Trait后的打印格式，比一般的打印信息更丰富；
（2）&‘static str 是Rust另一个折磨人的东西：生命周期。简单地说，Rust中每个引用都必须有一个生命周期，目的是当代码涉嫌制造悬垂指针的时候，Rust能够通过各个引用的生命周期标记，来作关系检查，杜绝悬垂指针。'static 是全局生命周期的意思。我也是刚刚接触这块内容，还在学习中。

**Comment by Bordertrail** [comment]
第四题也可以这样改
```
fn main() {
    define_x();
}

fn define_x() {
    let x = "hello";
    println!("{}, world", x);
}

**Reply by Dobedoooo** [reply]
我就是这么干的😁

**Reply by urnotsukilee** [reply]
我也是这么干的哈哈哈

**Comment by runningwa** [comment]
第八题加个mut

fn main() {
    let (mut x, y) = (1, 2);
    x += 2;

    assert_eq!(x, 3);
    assert_eq!(y, 2);
}

**Comment by rustin-huang** [comment]
感谢:D

**Comment by zhaoyuanmeng** [comment]
打卡

**Comment by maoxinkubi** [comment]
请问第三题怎么做

**Reply by PJ-568** [reply]
https://github.com/sunface/rust-by-practice/blob/master/solutions/variables.md

**Reply by livisun** [reply]
我把作用域给注释掉了
fn main() {
    let x: i32 = 10;
    // {
    let y: i32 = 5;
    println!("x 的值是 {}, y 的值是 {}", x, y);
    // }
    println!("x 的值是 {}, y 的值是 {}", x, y);
}

**Comment by ricardo1789** [comment]
找不到答案的 bro，看这里: https://github.com/sunface/rust-by-practice/blob/master/solutions/variables.md

**Comment by cmlark** [comment]
~~~rust
fn main() {
    let (x, y);
    (x,..) = (3, 4);
    [.., y] = [1, 2];
    // 填空，让代码工作
    assert_eq!([x,y], [3,2]);
} 
~~~

**虽然自己做对这道题，但我还是不理解这个解构式赋值，百度上也没查到什么有用的信息，请问何为“解构”呢，在赋值语句哪里中括号“[]”和小括号“（）”有什么区别吗，“...”又是做什么用的呢？**

**Reply by purple97** [reply]
1、是两个点“..”
2、解构 是把等号右边的内容赋值给左边，并且一一对应
3、两个点的意思类似 忽略， 忽略后面(x,..)，忽略前面(..,y)；

**Comment by Ezreal09** [comment]
学习一下

**Comment by axzed** [comment]
mark~

**Comment by zhusihan-python** [comment]
Unused variables 第七个已经不会报warning了，只有no output

**Comment by somecodework** [comment]
Done!

**Comment by worldstearly** [comment]
Homework done

**Comment by twandylue** [comment]
Today homework done!

**Comment by SkyeMoonchy** [comment]
homework done!

**Comment by axzed** [comment]
2022-12-17 mark

**Comment by xrr2016** [comment]
2022/12/20 Finished

**Comment by Bruce012356** [comment]
home work done.

**Comment by imguoguo** [comment]
打卡！

**Comment by nonopointer** [comment]
#[allow(unused_variables)] TNND玩阴滴是吧 (拔手雷.jpg)

**Comment by blkcor** [comment]
2023.1.2  学习rust打卡

**Reply by Kori000** [reply]
done

**Comment by cddsgtc** [comment]
pass

**Comment by imsellbaox** [comment]
mark

**Comment by suiang** [comment]
```rust
fn main() {
    let x = define_x();
    println!("{:?}, world", x); // 这里用{:?} 就会显示 ` "hello", world`
}

fn define_x() -> &'static str {
    let x = "hello";
    x
}

```

**Reply by suiang** [reply]
`println!("{:?}, world", x);`  //直接用 `{}`即可

**Comment by dlcnx** [comment]
mark

**Comment by SeddonShen** [comment]
打卡

**Comment by Non-PlayerC** [comment]
打卡，发现解构式赋值在rust中竟然还可以

```
let msg = Message::Move { x: 1, y: 2 };

match msg {
    Message::Quit => println!("Quit"),
    Message::Move { x, y } => {
        println!("Move to ({}, {})", x, y);
    }
    Message::Write(s) => {
        println!("Write message: {}", s);
    }
}
```

**Reply by Non-PlayerC** [reply]
在 Rust 中，:: 运算符用于访问某个类型或模块中的项（函数、变量、常量、类型等）

**Comment by Non-PlayerC** [comment]
备注: 所有的代码都是可以直接在浏览器修改运行的，体验极佳🙌

**Comment by tutdahai** [comment]
打卡

**Comment by CarryRee** [comment]
支持了

**Comment by mrhaoqi** [comment]
绑定 解构 遮蔽

**Comment by Mikachu2333** [comment]
第8题对初学者比如我来说不太好，因为在此时还没学到元组类型……

**Comment by 1226085293** [comment]
完成打开

**Comment by Chenleely** [comment]
done 9/9，打卡

**Comment by cz012273** [comment]
第4题不会，第7题没用答案中的方法（在后面加一个assert_eq!或println!）。

**Comment by TinyProbe** [comment]
4 is ambiguous. what does it want? and what exacly i can touch which part.

**Comment by lxian1988** [comment]
mark

**Comment by tushushu** [comment]
支持！

**Comment by x011223** [comment]
第6题也可以将 x 改成可变变量。
``` rust 
fn main() {
    let mut x: i32 = 1;
    x = 7;
    // 遮蔽且再次绑定
    let mut x = x; 
    x += 3;


    let y = 4;
    // 遮蔽
    let y = "I can also be bound to text!"; 
}
···

**Comment by lryain** [comment]
真6，书白买了！

**Comment by wulawulu** [comment]
mark一下，2023/07/10 0:03

**Comment by xbsheng** [comment]
打卡

**Comment by cyb-start** [comment]
练习很棒

**Comment by TURING132** [comment]
比想象中easy～

**Comment by cutimgboy** [comment]
mark一下

**Comment by qingfengmax03** [comment]
😄

**Comment by CynicalHeart** [comment]
mark here

**Comment by a1121611810** [comment]
基本都能答出来，除了那个需要多个方式，看答案是加#[allow(unused_variables)]这个就真不知道了

不过我是做前端的，先入主的观念，发现语法上确实好多异曲同工之处，理解上会稍稍容易点点

**Comment by Micro-ATP** [comment]
mark 一下

**Comment by 1259364843** [comment]
来了

**Comment by LandyZhu** [comment]
// 完形填空，让代码编译
fn main() {
    let x = 1;
    let x = x + 2;  // 与 let x += 2 有何区别呢？用 let x += 2 就说 x 不能修改；用 let x = x +2 就正常运行
    
    println!("x = {}", x); 
}

**Comment by hzzou** [comment]
做过一次

**Comment by jackrobet** [comment]
mark一下

**Comment by Muzych** [comment]
Mark 已完成

**Comment by OnlyAIGC** [comment]
已过

**Comment by thetaQ** [comment]
2023-11-28 mark

**Comment by Jeanstar96** [comment]
mark

**Comment by xiang-noodle** [comment]
done--2023/12/9

**Comment by yjshi2015** [comment]
为什么我在run时失败？
页面提示：Playground Communication: timeout。
控制台提示：访问 'https://play.rust-lang.org/evaluate.json'这个网站 CROS 跨域导致失败。

**Comment by yjshi2015** [comment]
rust学习第二天，困了……

**Comment by Hailaylin** [comment]
2023年12月15日 打卡，已过

**Comment by huahua132** [comment]
未使用的变量第二个答案是啥？

**Comment by keanedwards1** [comment]
sauce

**Comment by kieferEcho** [comment]
mark

**Comment by lostintime101** [comment]
thank you for this, really good exercises

**Comment by capybara0-0** [comment]
These are good exercises to understand the syntax,
Higly recommended.
Hands on practise

**Comment by thiru-01** [comment]
finished

**Comment by VessaliusOz** [comment]
mark

**Comment by ShafiqSadat** [comment]
Shafiq was here! 😉

**Comment by EthanOK** [comment]
Start!!!

**Comment by princemuel** [comment]
@princemuel was here 2024-02-07

**Comment by michahell** [comment]
I love how easy it is to learn Rust this way. Absolutely easy and effective exercises

**Comment by zhengkezhou1** [comment]
第一次

**Comment by vickyone0** [comment]
done

**Comment by felipe300** [comment]
- [x] Done

**Comment by ph03nix01** [comment]
mark

**Comment by sovansela** [comment]
--completed

**Comment by EmptyBlueBox** [comment]
checked!

**Comment by TDAkory** [comment]
mark finished

**Comment by reisenHSI** [comment]
很赞

**Comment by ALILIYES** [comment]
victory

**Comment by Classpi** [comment]
24/3/6

**Comment by LilMikey-CN** [comment]
mark一个

**Comment by ljzbtc** [comment]
check

**Comment by dontnet-wuenze** [comment]
谢谢，帮助很大

**Comment by Prajwalraut29** [comment]
awsome

**Comment by ItzAus6on** [comment]
mark一下

**Comment by thehaab** [comment]
complete

**Comment by erhszo** [comment]
```rust
fn main(){
    let status= define_status();
    println!("{}",status);
}

fn define_status()-> String{
    let x = "Completed.".to_string();
    x
}
```

**Comment by immadolf** [comment]
被自己笑死🤣

// 修复下面代码的错误并使用尽可能少的改变
fn main() {
    let x: i32 = 10;
    {
        let y: i32 = 5;
        println!("x 的值是 {}, y 的值是 {}", x, y);
    }
    println!("x 的值是 {}, y 的值是 {}", x, x); 
}

// 修复错误
fn main() {
    println!("{}, world", "hello"); 
}

fn define_x() {
    let x = "hello";
}

**Comment by C-RONALDO7V** [comment]
mark一下

**Comment by nerd4me** [comment]
mark @ 2024-04-30

**Comment by ZureJack** [comment]
没看到练习题阿

**Comment by anousonefs** [comment]
ຄັກໆ

**Comment by calc1f4r** [comment]
All correct✅

**Comment by sjunepark** [comment]
Done

**Comment by Kousaka-Mayuri** [comment]
Done

**Comment by shawnyplayz** [comment]
Great Shizz...

**Comment by chehongwei-git** [comment]
mark 一下

**Comment by winsonet** [comment]
有意思 :)

**Comment by ataliya-long** [comment]
学到了

**Comment by ospdz** [comment]
有的地方遮住了，又不能滚动，能修复一下就好了

**Comment by CryptoNinja0331** [comment]
Wonderful practice. 
This is extremely funny and helpful for beginners! Thanks.

**Comment by MoXins** [comment]
2024.7.9打卡第一天

**Comment by jkbody** [comment]
mark

**Comment by CryptoNinja0331** [comment]
Really helpful!

**Comment by AlkaidLu** [comment]
maek

**Comment by stoneson** [comment]
mark一下

**Comment by rocky-rose** [comment]
I accidentally oversimplified the solution to #4 :sweat_smile:.

```rust
// Fix the error with the use of define_x
fn main() {
    define_x(); 
}

fn define_x() {
    let x = "hello";
    println!("{}, world", x);
}
```

I'm practicing the two solutions that involve returning `x`.

**Comment by zxwitsme** [comment]
mark

**Comment by ScreechingGrapeLeviathan** [comment]
mark

**Comment by mtdxwb** [comment]
mark

**Comment by LiaBrio** [comment]
check

**Comment by chenmx00** [comment]
mark

**Comment by hoBabu1** [comment]
done

**Comment by LonelyFellas** [comment]
done and mark

**Comment by ovie-daveth** [comment]
I did great here!!!

**Comment by Sagecheni** [comment]
done

**Comment by RiverAi7z** [comment]
mark

**Comment by bbhpt** [comment]
mark

**Comment by gptshubham** [comment]
mark

**Comment by zhengkezhou1** [comment]
day1

**Comment by Glessy-nf** [comment]
mark 一下

**Comment by Paul-from-NJTech** [comment]
mark

**Comment by Abhinav-Singh10** [comment]
done

**Comment by UnclePhil1** [comment]
Completed

**Comment by bansal-harsh-2504** [comment]
great way to learn.

**Comment by Ali-Adel-Nour** [comment]
9/9

**Comment by mohamedelbalshy** [comment]
this is very helpful, thanks so much for this project.

**Comment by chttiu** [comment]
标记完成

**Comment by CamaroW** [comment]
题目的代码块最后一行总是只有1/3可见，不知道大家都这样还是只有我。比如第7题的最后一行注释“// compiler warning: unused variable: `x`”，就看不见，但是能选中然后粘贴出来看哈哈哈哈。

**Reply by chttiu** [reply]
> 题目的代码块最后一行总是只有1/3可见，不知道大家都这样还是只有我。比如第7题的最后一行注释“// compiler warning: unused variable: `x`”，就看不见，但是能选中然后粘贴出来看哈哈哈哈。

是的，有些代码块没办法滚动

**Comment by Tenax-599** [comment]
fn main() {
    let mut x: i32 = 1;
    x = 7;
    // 遮蔽且再次绑定
    let mut x = x; 
    x += 3;


    let y = 4;
    // 遮蔽
    let y = "I can also be bound to text!"; 
}
第六题我是这样修改的

**Comment by Thomas-YHS** [comment]
mark 2025.01.06

**Comment by Mao-ann** [comment]
想问问大佬为什么第四题这样改会显示超时啊？（以下是代码）
let x="hello";
fn main(){ println!("{},world!",x) }

**Reply by CamaroW** [reply]
那个 playground 好像有时候就是会超时（课程页面嵌入的和网站官网 https://play.rust-lang.org/?version=stable&mode=debug&edition=2021 都会），等几分钟，或者本地运行下试试。

**Comment by davelet** [comment]
第4题这样可以吗
```rust
fn main() {
    let x = define_x();
    println!("{}, world", x); 
}

fn define_x() -> String {
    let x = "hello";
    x.into()
}
```

**Comment by FernandezFigo105** [comment]
Figo was here.

**Comment by AHL8192** [comment]
yoo

**Comment by Luyoung0001** [comment]
mark！

**Comment by in-people** [comment]
fn main() {
    let x: i32 = 10;
    {
        let y: i32 = 5;  y是局部变量；
        println!("x 的值是 {}, y 的值是 {}", x, y);
    }
    println!("x 的值是 {}, y 的值是 {}", x, y);  在这个作用域中找不到y。
error[E0425]: cannot find value `y` in this scope

**Reply by in-people** [reply]
作用域是一个变量在程序中能够保持合法的范围

**Reply by in-people** [reply]
// 修复错误
fn main() {
    let x = "hello";
    define_x();
    println!("{}, world", x); 
}

fn define_x() {
    let x = "hello";
}

main 函数中的 x 是一个局部变量，值为 "hello"。
define_x 函数中的 x 是另一个局部变量，虽然名字相同，但它与 main 函数中的 x 没有任何关系。

**Reply by in-people** [reply]
总结来说，assert_eq!(x, 12); 是一种用于确保 x 的值等于 12 的断言工具，常用于测试和调试
如果x等于12，程序正常执行；否则，程序会触发一个 panic（崩溃），并输出一条错误消息，指出两个值不相等。

**Reply by in-people** [reply]
error[E0384]: cannot assign twice to immutable variable `x`
 --> src/main.rs:7:5
  |
6 |     let x = x; 
  |         - first assignment to `x`
7 |     x += 3;
  |     ^^^^^^ cannot assign twice to immutable variable 这是什么意思？

**Reply by in-people** [reply]
感觉rust特别规范，每个错误编译器都会明确指出。

**Comment by Equinocciox** [comment]
mark一下

**Comment by ZaneLeo1** [comment]
mark一下

**Comment by jinitaimei-jige-114514** [comment]
mark

**Comment by eyenov** [comment]
头一次做题，记录一下

**Comment by johannblaser** [comment]
how to show up the warning when running this: 
fn main() {
    let x = 1; 
}

**Reply by johannblaser** [reply]
i solved it myself

#![deny(warnings)]
fn main() {
    let x = 1; 
}

// Warning: unused variable: `x`

**Comment by xvzhnang** [comment]
mark一下

**Comment by Seine-yh** [comment]
Good!

**Comment by VWWVVW** [comment]
很好的教程与练习，使我的电脑生锈。
另外有一个可能的bug：
练习给的代码前面都有一个空行，最后一行总是被遮住一半，换了个浏览器也一样，不知道是不是只有我有这个问题，虽然可以通过敲回车解决，但强迫症还是受不了 : )

**Comment by wling-art** [comment]
mark - 2025.4.16

**Comment by obhor** [comment]
omkar was here

**Comment by ebdurrehm** [comment]
done
